rir/src/compiler/opt/cleanup.cpp

```diff
@@ -130,20 +130,24 @@ class TheCleanup {
         std::unordered_map<BB*, BB*> toDel;
         Visitor::run(function->entry, [&](BB* bb) {
             // Remove unnecessary splits
-            if (bb->isJmp() && cfg.hasSinglePred(bb->next0) &&
+            if (bb->isJmp() && cfg.hasSinglePred(bb->next0)) {
                 /* this condition keeps graph in split-edge: */
```

Comment:
From what I understand, this second condition is conservative. I mean *bb* could have 2 predecessors, each just being a jump and I think that does not mean the edge can not be removed without breaking the condition. If I am wrong ok, otherwise a comment clarifying would help.

Beyond that, it would be great to have both `ifs` following the same order for the conditions:
1. isJmp() && 
2. hasSinglePred()

Then, maybe 1 && 2 could be added as a method to cfg. I am not sure the proper name for that, maybe isIntermediary?

Reply:
now you confused myself. while the condition is conservative regarding whether the graph is still edge-split, it is of course a necessary condition for the question whether I can dump instructions into the block. I mean, if not `cfg.hasSinglePred(bb->next0)` that basically means:
```
   bb      other
    \    /
      next0
```

so if i merge `bb` with `next0`, or if i dump instructions into `next0`, in both cases bad things will happen.

so the first `if` is a necessary and sufficient condition wheter I can flow instructions from `bb` to `next0`, and the second if is a neccessary and sufficient condition on the question whether i can merget `bb` and `next0` into one, without breaking edge-split.

Reply:
ignore what I wrote, I think it's still not quite accurate